btas/generic/tuck_cp_als.h

```diff
@@ -0,0 +1,500 @@
+    /// \param[in] rank current rank of the decomposotion
+    /// \param[in] converge_test ALS satisfactory condition checker.
+    /// \param[in] dir not used in this function
+    /// \param[in] max_als maximum number of ALS iterations
+    /// \param[in] calculate_epsilon should epsilon be returned, disregarded if ConvClass = FitCheck
+    /// \param[in, out] epsilon in: a double value is disregarded.
+    /// out: if ConvClass = FitCheck || \c calculate_epsilon the 2-norm tensor
+    /// error of the CP approximation else not modified.
+    /// \param[in] fast_pI Should ALS use a faster version of pseudoinverse?
+    void ALS(ind_t rank, ConvClass &converge_test, bool dir, ind_t max_als, bool calculate_epsilon, double &epsilon,
+             bool &fast_pI) {
+      size_t count = 0;
+
+      if(tucker_factors.empty()) {
+        make_tucker_factors(tensor_ref, this->tcut_tucker, tucker_factors, false);
+        core_tensor = tensor_ref;
+        transform_tucker(true, core_tensor, tucker_factors);
+      }
+
+      if(AtA.empty()) {
+        AtA = std::vector<Tensor>(ndim);
+        transformed_A = std::vector<Tensor>(ndim);
+      }
+      auto ptr_A = A.begin(), ptr_T = tucker_factors.begin(),
+           ptr_AtA = AtA.begin(), ptr_tran = transformed_A.begin();
+      for (size_t i = 0; i < ndim; ++i,++ptr_A, ++ptr_T, ++ptr_AtA, ++ptr_tran) {
+        auto &a_mat = A[i];
+        *ptr_AtA = Tensor();
+        contract(1.0, *ptr_A, {1, 2}, *ptr_A, {1, 3}, 0.0, *ptr_AtA, {2, 3});
+        Tensor trans;
+        *ptr_tran = Tensor();
+        contract(1.0, *ptr_T, {1,2}, *ptr_A, {2,3}, 0.0, *ptr_tran, {1,3});
+      }
+
+      helper = RALSHelper<Tensor>(A);
+      const auto s0 = 1.0;
+      std::vector<double> lambda(ndim, 1.0);
+      const auto alpha = 0.8;
+
+      // Until either the initial guess is converged or it runs out of iterations
+      // update the factor matrices with or without Khatri-Rao product
+      // intermediate
+      bool is_converged = false;
+      bool matlab = fast_pI;
+      while (count < max_als && !is_converged) {
+        count++;
+        this->num_ALS++;
+        for (size_t i = 0; i < ndim; i++) {
+          this->direct(i, rank, fast_pI, matlab, converge_test, lambda[i]);
+          // Compute the value s after normalizing the columns
+          auto & ai = A[i];
+          this->s = helper(i, ai);
+          // recompute lambda
+          lambda[i] = (lambda[i] * (this->s * this->s) / (s0 * s0)) * alpha + (1 - alpha) * lambda[i];
+          contract(1.0, tucker_factors[i], {1,2}, ai, {2,3},0.0, transformed_A[i], {1,3});
+          contract(1.0, ai, {1,2}, ai, {1,3}, 0.0, AtA[i], {2,3});
+        }
+        is_converged = converge_test(A);
+      }
+
+      // Checks loss function if required
+      detail::get_fit(converge_test, epsilon);
+      epsilon = 1 - epsilon;
+      // Checks loss function if required
+      if (calculate_epsilon && typeid(converge_test) != typeid(btas::FitCheck<Tensor>)) {
```

Comment:
ditto

Reply:
`typeid` is unnecessary and has been removed